File: HealthMindControl/src/healthmind_control/kafka.py
import asyncio
import json
from datetime import datetime
from typing import Any

from confluent_kafka import Consumer, ConsumerGroupTopicPartitions, KafkaException, Producer, TopicPartition
from confluent_kafka.admin import AdminClient

from .config import Settings


BUSINESS_TOPICS = {
    "nutrition-capture-ready",
    "nutrition-analysis-completed",
    "nutrition-analysis-failed",
}


class KafkaService:
    def __init__(self, cfg: Settings):
        self.cfg = cfg
        base = {"bootstrap.servers": cfg.kafka_bootstrap_servers, "socket.timeout.ms": 5000}
        self.admin = AdminClient(base)
        self.producer = Producer(base)

    async def metadata(self) -> dict[str, Any]:
        return await asyncio.to_thread(self._metadata)

    def _metadata(self) -> dict[str, Any]:
        md = self.admin.list_topics(timeout=5)
        brokers = [{"id": b.id, "host": b.host, "port": b.port} for b in md.brokers.values()]
        topics = []
        consumer = Consumer({"bootstrap.servers": self.cfg.kafka_bootstrap_servers, "group.id": "healthmind-control-metadata", "enable.auto.commit": False})
        for name, topic in sorted(md.topics.items()):
            if name.startswith("__"):
                continue
            parts = []
            for pid, part in sorted(topic.partitions.items()):
                try:
                    low, high = consumer.get_watermark_offsets(TopicPartition(name, pid), timeout=2)
                except Exception:
                    low, high = None, None
                parts.append({
                    "partition": pid, "leader": part.leader,
                    "replicas": list(part.replicas), "isr": list(part.isrs),
                    "earliest_offset": low, "latest_offset": high,
                    "healthy": part.leader >= 0 and set(part.replicas) == set(part.isrs),
                })
            topics.append({"name": name, "error": str(topic.error) if topic.error else None, "partitions": parts})
        consumer.close()
        return {"brokers": brokers, "topics": topics}

    async def groups(self) -> list[dict[str, Any]]:
        return await asyncio.to_thread(self._groups)

    def _groups(self) -> list[dict[str, Any]]:
        result = self.admin.list_consumer_groups(request_timeout=5).result(6)
        groups = []
        for listing in result.valid:
            item = {"group_id": listing.group_id, "state": str(listing.state), "simple": listing.is_simple_consumer_group, "lag": None}
            try:
                offsets = self.admin.list_consumer_group_offsets([ConsumerGroupTopicPartitions(listing.group_id)], request_timeout=5)[listing.group_id].result(6)
                md = self.admin.list_topics(timeout=5)
                lag = 0
                consumer = Consumer({"bootstrap.servers": self.cfg.kafka_bootstrap_servers, "group.id": "healthmind-control-lag", "enable.auto.commit": False})
                for tp in offsets.topic_partitions:
                    if tp.offset >= 0 and tp.topic in md.topics:
                        _, high = consumer.get_watermark_offsets(TopicPartition(tp.topic, tp.partition), timeout=2)
                        lag += max(0, high - tp.offset)
                consumer.close(); item["lag"] = lag
            except Exception:
                pass
            groups.append(item)
        return groups

    async def messages(
        self, topic: str, partition: int = 0, start: str = "latest", offset: int | None = None,
        limit: int = 50, contains: str | None = None,
    ) -> list[dict[str, Any]]:
        return await asyncio.to_thread(self._messages, topic, partition, start, offset, min(limit, 100), contains)

    def _messages(self, topic: str, partition: int, start: str, offset: int | None, limit: int, contains: str | None):
        consumer = Consumer({
            "bootstrap.servers": self.cfg.kafka_bootstrap_servers,
            "group.id": f"healthmind-control-diagnostic-{datetime.now().timestamp()}",
            "enable.auto.commit": False,
            "enable.auto.offset.store": False,
            "auto.offset.reset": "earliest",
        })
        try:
            tp = TopicPartition(topic, partition)
            low, high = consumer.get_watermark_offsets(tp, timeout=5)
            target = offset if offset is not None else (low if start == "earliest" else max(low, high - limit))
            consumer.assign([TopicPartition(topic, partition, target)])
            output = []
            empty = 0
            while len(output) < limit and empty < 3:
                msg = consumer.poll(1)
                if msg is None:
                    empty += 1
                    continue
                if msg.error():
                    if msg.error().code() == -191:
                        break
                    raise KafkaException(msg.error())
                raw = msg.value().decode("utf-8", "replace") if msg.value() else ""
                if contains and contains not in raw and contains not in (msg.key() or b"").decode("utf-8", "replace"):
                    continue
                try:
                    payload = json.loads(raw)
                except json.JSONDecodeError:
                    payload = raw
                output.append({
                    "topic": msg.topic(), "partition": msg.partition(), "offset": msg.offset(),
                    "timestamp": msg.timestamp()[1],
                    "key": (msg.key() or b"").decode("utf-8", "replace"),
                    "headers": dict(msg.headers() or []), "payload": payload,
                })
            return output
        finally:
            consumer.close()

    def validate_event(self, topic: str, payload: Any) -> list[str]:
        warnings = []
        if topic in BUSINESS_TOPICS:
            required = {"event_id", "event_type", "producer", "trace_id", "schema_version", "payload"}
            if not isinstance(payload, dict):
                warnings.append("业务 topic 的 payload 应为 JSON 对象")
            else:
                missing = sorted(required - payload.keys())
                if missing:
                    warnings.append("缺少业务事件字段: " + ", ".join(missing))
        return warnings

    async def produce(self, topic: str, key: str | None, payload: Any, headers: dict[str, str], partition: int | None):
        return await asyncio.to_thread(self._produce, topic, key, payload, headers, partition)

    def _produce(self, topic: str, key: str | None, payload: Any, headers: dict[str, str], partition: int | None):
        delivered: dict[str, Any] = {}
        def callback(error, message):
            if error:
                delivered["error"] = str(error)
            else:
                delivered.update(topic=message.topic(), partition=message.partition(), offset=message.offset())
        kwargs = {"topic": topic, "key": key, "value": json.dumps(payload, ensure_ascii=False), "headers": list(headers.items()), "callback": callback}
        if partition is not None:
            kwargs["partition"] = partition
        self.producer.produce(**kwargs)
        remaining = self.producer.flush(10)
        if remaining or delivered.get("error"):
            raise RuntimeError(delivered.get("error") or f"{remaining} message(s) undelivered")
        return delivered

File: HealthMindControl/src/healthmind_control/static/js/views/kafka.js
// Kafka 管理视图
import { esc, fmtTs, getJson, postJson, toast, $, openDrawer } from "../main.js";
import { renderJsonTreeEl } from "../json-tree.js";
import { startMutation } from "../flow.js";

let topicsMeta = { brokers: [], topics: [] };
let groupsData = [];
let timerM = null;

const BIZ = ["nutrition-capture-ready", "nutrition-analysis-completed", "nutrition-analysis-failed"];

function partHealth(topic, p) {
  const isOk = !!(p.healthy);
  return '<span class="badge ' + (isOk ? "b-ok" : "b-err") + '" title="leader=' + p.leader + " replicas=[" + (p.replicas || []).join(",")
    + "] isr=[" + (p.isrs || []).join(",") + "]" + (p.error ? " err:" + p.error : "") + '">p' + p.partition + (isOk ? "" : " ⚠") + "</span>";
}

function renderTopics() {
  const box = document.getElementById("k-topics");
  if (!topicsMeta.topics || !topicsMeta.topics.length) { box.innerHTML = '<div class="muted">无法获取 topic 列表</div>'; return; }
  const html = topicsMeta.topics.map(function (t) {
    const unhealthy = (t.partitions || []).filter(function (p) { return !p.healthy; });
    const parts = (t.partitions || []).map(function (p) {
      return '<div style="display:flex;justify-content:space-between;gap:6px;padding:2px 0">'
        + '<span class="badge ' + (p.healthy ? "b-ok" : "b-err") + '">p' + p.partition + "</span>"
        + '<span class="small muted mono">earliest ' + (p.earliest_offset ?? "-") + " · latest " + (p.latest_offset ?? "-") + "</span></div>";
    }).join("");
    return '<div style="padding:8px 0;border-bottom:1px solid var(--line-soft)">'
      + '<div class="row" style="gap:8px;flex-wrap:wrap"><b class="mono">' + esc(t.name) + "</b>"
      + (BIZ.includes(t.name) ? '<span class="badge b-accent">业务</span>' : "")
      + (unhealthy.length ? '<span class="badge b-err">异常分区 ' + unhealthy.length + "</span>" : '<span class="badge b-ok">健康</span>')
      + "</div><div>" + parts + "</div></div>";
  }).join("");
  box.innerHTML = '<div class="row small muted" style="margin-bottom:4px">brokers: ' + (topicsMeta.brokers || []).map(function (b) { return esc(b.host + ":" + b.port); }).join(", ") + "</div>" + html;
  fillSelects();
}

function fillSelects() {
  const topicSel = document.getElementById("m-topic");
  const current = topicSel.value;
  topicSel.innerHTML = topicsMeta.topics.map(function (t) { return '<option value="' + esc(t.name) + '">' + esc(t.name) + "</option>"; }).join("");
  if (current && topicsMeta.topics.some(function (t) { return t.name === current; })) topicSel.value = current;
  topicChanged();
  document.getElementById("topic-list").innerHTML = topicsMeta.topics.map(function (t) { return '<option value="' + esc(t.name) + '">'; }).join("");
}

function topicChanged() {
  const t = topicsMeta.topics.find(function (x) { return x.name === document.getElementById("m-topic").value; });
  const partSel = document.getElementById("m-partition");
  if (!t) { partSel.innerHTML = ""; return; }
  partSel.innerHTML = t.partitions.map(function (p) { return '<option value="' + p.partition + '">分区 ' + p.partition + "</option>"; }).join("");
  const p = t.partitions[0] || {};
  document.getElementById("m-range").textContent = "earliest=" + (p.earliest_offset ?? "-") + " latest=" + (p.latest_offset ?? "-");
}

function renderGroups() {
  const box = document.getElementById("k-groups");
  groupsData = groupsData.filter(function (g) { return !String(g.group_id).startsWith("healthmind-control-"); });
  if (!groupsData.length) { box.innerHTML = '<div class="muted">无消费组</div>'; return; }
  box.innerHTML = groupsData.map(function (g) {
    const rows = (g.offsets || []).map(function (o) {
      return '<tr><td class="mono-cell">' + esc(o.topic) + "</td><td class=\"num\">" + o.partition + "</td><td class=\"num\">" + o.offset + "</td><td class=\"num\">" + (o.high ?? "-") + "</td><td class=\"num\" style=\"color:" + ((o.lag || 0) > 0 ? "var(--warn)" : "var(--ok)") + '">' + (o.lag ?? "-") + "</td></tr>";
    }).join("");
    const members = (g.members || []).map(function (m) {
      const asg = (m.assignments || []).map(function (a) { return esc(a.topic) + "/" + a.partition; }).join(" ");
      return '<div class="small muted mono" title="' + esc(m.client_id || "") + " @ " + esc(m.host || "") + '">' + esc(String(m.member_id || "").slice(0, 22)) + (asg ? " · " + asg : "") + "</div>";
    }).join("");
    return '<details style="padding:8px 0;border-bottom:1px solid var(--line-soft)">'
      + '<summary class="row" style="gap:8px;cursor:pointer"><b class="mono">' + esc(g.group_id) + "</b>"
      + '<span class="badge">' + esc(g.state) + "</span>"
      + '<span class="badge ' + ((g.lag || 0) > 0 ? "b-warn" : "b-ok") + '">lag ' + (g.lag || 0) + "</span>"
      + '<span class="small muted">成员 ' + (g.members || []).length + "</span></summary>"
      + '<div style="margin:6px 0">' + (members || '<span class="muted small">无在线成员</span>') + "</div>"
      + '<div class="tbl-wrap"><table class="tbl"><thead><tr><th>topic</th><th style="text-align:right">分区</th><th style="text-align:right">offset</th><th style="text-align:right">high</th><th style="text-align:right">lag</th></tr></thead><tbody>' + (rows || '<tr><td colspan="5" class="empty">无 offset</td></tr>') + "</tbody></table></div>"
      + "</details>";
  }).join("");
}

async function loadTopics(force) {
  try { topicsMeta = await getJson("/api/kafka/topics" + (force ? "?force=true" : "")); renderTopics(); }
  catch (e) { document.getElementById("k-topics").innerHTML = '<div class="muted">加载失败：' + esc(e.message) + "</div>"; }
}
async function loadGroups() {
  try { groupsData = await getJson("/api/kafka/groups"); renderGroups(); }
  catch (e) { document.getElementById("k-groups").innerHTML = '<div class="muted">加载失败：' + esc(e.message) + "</div>"; }
}

// ---------------- 消息查看 ----------------
function messageRow(m, idx) {
  const preview = typeof m.payload === "string" ? m.payload.slice(0, 90) : JSON.stringify(m.payload).slice(0, 90);
  return '<tr style="cursor:pointer" data-msg="' + idx + '">'
    + '<td class="time-cell">' + fmtTs(m.timestamp) + "</td>"
    + '<td class="num">' + m.offset + "</td>"
    + '<td class="mono-cell ellip" style="max-width:150px" title="' + esc(m.key || "") + '">' + esc(m.key || "-") + "</td>"
    + '<td class="num">' + Object.keys(m.headers || {}).length + "</td>"
    + '<td class="mono-cell ellip" style="max-width:300px">' + esc(preview) + "</td></tr>";
}

async function readMessages() {
  const params = { topic: document.getElementById("m-topic").value, partition: document.getElementById("m-partition").value };
  const start = document.getElementById("m-start").value;
  params.start = start;
  if (start === "offset") params.offset = document.getElementById("m-offset").value;
  if (start === "time") {
    const v = document.getElementById("m-time").value;
    if (!v) { toast("请选择时间", "warn"); return; }
    params.time_ms = String(new Date(v).getTime());
  }
  ["key", "event", "trace", "type", "contains"].forEach(function (k) {
    const v = document.getElementById("m-" + (k === "event" ? "event" : k === "type" ? "type" : k)).value;
    if (v) params[k === "event" ? "event_id" : k === "type" ? "event_type" : k] = v;
  });
  const qs = new URLSearchParams(params).toString();
  document.getElementById("m-status").textContent = "读取中…（扫描上限 5 万条）";
  try {
    const list = await getJson("/api/kafka/messages?" + qs);
    const truncated = list.some(function (m) { return m.truncated_by_scan; });
    const real = list.filter(function (m) { return !m.truncated_by_scan; });
    const body = document.getElementById("m-messages");
    if (!real.length) { body.innerHTML = '<div class="muted">该位置与过滤条件下没有消息' + (truncated ? "（已到扫描上限）" : "") + "</div>"; }
    else {
      body.innerHTML = '<div class="tbl-wrap"><table class="tbl"><thead><tr><th>时间</th><th style="text-align:right">offset</th><th>key</th><th style="text-align:right">headers</th><th>payload 预览</th></tr></thead><tbody>'
        + real.map(function (m, i) { return messageRow(m, i); }).join("") + "</tbody></table></div>"
        + (truncated ? '<div class="small muted" style="margin-top:4px">已达扫描上限，还有更多匹配消息未显示</div>' : "");
      const rows = Array.from(body.querySelectorAll("tr[data-msg]"));
      rows.forEach(function (tr) {
        tr.addEventListener("click", function () {
          const msg = real[Number(tr.dataset.msg)];
          showMessage(msg);
        });
      });
    }
    document.getElementById("m-status").textContent = "返回 " + real.length + " 条" + (truncated ? "（截断）" : "");
  } catch (e) { toast("读取失败：" + e.message, "err"); document.getElementById("m-status").textContent = ""; }
}

function showMessage(m) {
  const meta = '<div class="msg-meta" style="margin-bottom:10px">'
    + "<div><b>Topic</b> " + esc(m.topic) + "</div>"
    + "<div><b>分区/Offset</b> " + m.partition + " / " + m.offset + "</div>"
    + "<div><b>时间</b> " + fmtTs(m.timestamp) + "</div>"
    + "<div><b>Key</b> " + esc(m.key || "-") + "</div>"
    + "</div>";
  const headers = Object.keys(m.headers || {}).length
    ? '<h3 style="font-size:13px;margin:8px 0 4px">Headers</h3><div class="msg-meta">' + Object.entries(m.headers).map(function (h) { return "<div><b>" + esc(h[0]) + "</b> " + esc(h[1]) + "</div>"; }).join("") + "</div>"
    : "";
  const payload = typeof m.payload === "string" ? m.payload : m.payload;
  const wrap = document.createElement("div");
  wrap.innerHTML = meta + headers;
  const title = document.createElement("div");
  title.innerHTML = '<h3 style="font-size:13px;margin:8px 0 4px">Payload</h3>';
  wrap.appendChild(title);
  wrap.appendChild(renderJsonTreeEl(payload));
  openDrawer("消息 " + m.topic + "/" + m.partition + "@" + m.offset, wrap);
}

// ---------------- 受控生产 ----------------
function bizTemplate() {
  const topic = document.getElementById("p-topic").value.trim();
  if (!BIZ.includes(topic)) { toast("仅三个业务 topic 提供模板", "warn"); return; }
  const now = new Date();
  const event = {
    event_id: crypto.randomUUID(),
    event_type: { "nutrition-capture-ready": "nutrition.capture.ready.v1", "nutrition-analysis-completed": "nutrition.analysis.completed.v1", "nutrition-analysis-failed": "nutrition.analysis.failed.v1" }[topic],
    occurred_at: now.toISOString(),
    producer: topic === "nutrition-capture-ready" ? "NutriMemo" : "HealthMind",
    trace_id: crypto.randomUUID(),
    subject_id: null,
    aggregate_type: "meal",
    aggregate_id: "",
    schema_version: "1.0",
    payload: {},
  };
  document.getElementById("p-payload").value = JSON.stringify(event, null, 2);
}

function headersObject() {
  const out = {};
  document.querySelectorAll("#p-headers .h-row").forEach(function (row) {
    const k = row.querySelector(".hk").value.trim();
    const v = row.querySelector(".hv").value;
    if (k) out[k] = v;
  });
  return out;
}
function addHeaderRow(k, v) {
  const box = document.getElementById("p-headers");
  const row = document.createElement("div");
  row.className = "row h-row";
  row.style.gap = "4px";
  row.style.marginBottom = "4px";
  row.innerHTML = '<input type="text" class="hk" placeholder="key" value="' + esc(k || "") + '" style="flex:1">'
    + '<input type="text" class="hv" placeholder="value" value="' + esc(v || "") + '" style="flex:1">'
    + '<button class="btn sm danger" type="button">✕</button>';
  row.querySelector("button").addEventListener("click", function () { row.remove(); });
  box.appendChild(row);
}

function previewProduce() {
  const topic = document.getElementById("p-topic").value.trim();
  if (!topic) { toast("请填写 topic", "warn"); return; }
  let payload;
  try { payload = JSON.parse(document.getElementById("p-payload").value); }
  catch { toast("payload 不是合法 JSON", "err"); return; }
  const key = document.getElementById("p-key").value.trim() || null;
  const partitionRaw = document.getElementById("p-partition").value;
  const partition = partitionRaw === "" ? null : Number(partitionRaw);
  const reason = document.getElementById("p-reason").value.trim();
  const headers = headersObject();
  const target = document.getElementById("p-preview-box");
  startMutation(target, {
    previewUrl: "/api/kafka/produce/preview",
    executeUrl: "/api/kafka/produce/execute",
    title: "生产消息",
    body: function () { return { topic: topic, key: key, partition: partition, headers: headers, payload: payload, reason: reason }; },
    onDone: function (result) {
      if (result && result.offset !== undefined) {
        document.getElementById("p-preview-box").innerHTML = '<div class="row"><span class="badge b-ok">delivered</span>'
          + '<span class="mono">partition=' + result.partition + " offset=" + result.offset + "</span></div>";
      }
    },
  });
}

function init() {
  document.getElementById("k-reload").addEventListener("click", function () { loadTopics(true); });
  document.getElementById("g-reload").addEventListener("click", loadGroups);
  document.getElementById("m-topic").addEventListener("change", topicChanged);
  document.getElementById("m-start").addEventListener("change", function () {
    const s = document.getElementById("m-start").value;
    document.getElementById("m-offset-field").hidden = s !== "offset";
    document.getElementById("m-time-field").hidden = s !== "time";
  });
  document.getElementById("m-read").addEventListener("click", readMessages);
  document.getElementById("m-auto").addEventListener("change", function () {
    if (this.checked) { readMessages(); timerM = setInterval(function () { if (!document.hidden) readMessages(); }, 15000); }
    else if (timerM) { clearInterval(timerM); timerM = null; }
  });
  document.getElementById("p-hadd").addEventListener("click", function () {
    const hk = document.getElementById("p-hk"); const hv = document.getElementById("p-hv");
    addHeaderRow(hk.value, hv.value); hk.value = ""; hv.value = "";
  });
  document.getElementById("p-template").addEventListener("click", bizTemplate);
  document.getElementById("p-pretty").addEventListener("click", function () {
    try { document.getElementById("p-payload").value = JSON.stringify(JSON.parse(document.getElementById("p-payload").value), null, 2); }
    catch { toast("不是合法 JSON", "err"); }
  });
  document.getElementById("p-preview").addEventListener("click", previewProduce);
  loadTopics(false);
  loadGroups();
}
init();
